### DISCUSSION TITLE
Suggestions for the code generator

### DISCUSSION THREAD

**Discussion by osaxma** [serverpod/serverpod#110] [Ideas]
I didn't wanna lose my thoughts as I'm tinkering with the package so I'm compiling some notes here quickly that can be promoted to separate issues later, if necessary:

### - Parse `double` explicitly 

**Current**: [serverpod_cli/bin/generator/class_generator_dart.dart#L600](https://github.com/serverpod/serverpod/blob/f2379776ddfcbc49e1017a571f235ba9292d28d1/tools/serverpod_cli/bin/generator/class_generator_dart.dart#L600)

**Suggestion**: Use `_data["some_double"].toDouble()` or `_data["some_double"]?.toDouble()` for nullable.

**Reason**: double can be converted to int when encoded in Web (`1.0` encoded as `1` that's later decoded as `int` causing a runtime error for type mismatch) (see: https://github.com/dart-lang/sdk/issues/45160) (got bitten there before)

### - Missing equality (i.e. `operator==` & `hashCode`), `copyWith`, and `toString` in generated code

Currently the generated code does not include any of the three. They are useful methods/getters which are commonly used and also are typically generated (optional) by popular data class generators such as Freezed and Dart "Data Class Generator" (VS Code Extension).

**Example**: see https://github.com/osaxma/dfs/blob/87e1f65b3bb63f1f0c1d05baf80f3e404a387816/samples/data_class_samples/output_data_example.dart#L14

###  - Any reason why the generator is not utilizing the [`code_builder`](https://pub.dev/packages/code_builder) package?

I believe it helps to generate dart code in a type safe manner that's easy to refractor instead of concatenating strings which can be error prone and hard to refractor. Though, I don't think this is an issue atm since the generated code for serializable classes isn't very complicated. 

--- 

I hope the suggestions are found useful and I hope I'll be able to contribute more in the future. 

Thank you.

**Comment by ondbyte** [comment]
I recently came across serverpod, I'm with you on the 2 and 3 points, using code_builder is a very good practice

**Comment by vlidholt** [comment]
Great suggestions @osaxma. I will definitely look into that.

Only reason I haven't used code_builder is because I haven't found it. Looks like a much more clean solution for sure.
